File: src/schema/diff.ts
// Structural schema comparison — detect changes between stored and code schemas

import type { CompiledSchema, StoredSchema, StoredColumnDef, CompiledField } from "../types.ts";

export interface SchemaChange {
  type: "added" | "removed" | "typeChanged" | "requireChanged";
  field: string;
  oldType?: string;
  newType?: string;
}

export function compiledToStored(schema: CompiledSchema): StoredSchema {
  return {
    columns: schema.fields.map((f) => ({
      name: f.name,
      type: f.kind,
      required: f.required || undefined,
      unique: f.unique || undefined,
    })),
  };
}

export function diffSchemas(
  stored: StoredSchema,
  current: CompiledSchema,
): SchemaChange[] {
  const changes: SchemaChange[] = [];

  const storedMap = new Map<string, StoredColumnDef>();
  for (const col of stored.columns) {
    storedMap.set(col.name, col);
  }

  const currentMap = new Map<string, CompiledField>();
  for (const field of current.fields) {
    currentMap.set(field.name, field);
  }

  // Check for removed fields
  for (const [name] of storedMap) {
    if (!currentMap.has(name)) {
      changes.push({ type: "removed", field: name });
    }
  }

  // Check for added or changed fields
  for (const [name, field] of currentMap) {
    const old = storedMap.get(name);
    if (!old) {
      changes.push({ type: "added", field: name });
    } else if (old.type !== field.kind) {
      changes.push({
        type: "typeChanged",
        field: name,
        oldType: old.type,
        newType: field.kind,
      });
    }
  }

  return changes;
}

export function schemasEqual(stored: StoredSchema, current: CompiledSchema): boolean {
  if (stored.columns.length !== current.fields.length) return false;

  for (let i = 0; i < stored.columns.length; i++) {
    const s = stored.columns[i];
    const c = current.fields[i];
    if (s.name !== c.name || s.type !== c.kind) return false;
  }

  return true;
}

export function validateMigration(
  changes: SchemaChange[],
  migrations: { version: number; rename?: Record<string, string>; transform?: unknown }[],
  targetVersion: number,
): string[] {
  const errors: string[] = [];
  const migration = migrations.find((m) => m.version === targetVersion);
  const renames = migration?.rename ?? {};
  const hasTransform = !!migration?.transform;

  // Renamed fields are expected to show as "removed old" + "added new"
  const renameOldNames = new Set(Object.keys(renames));
  const renameNewNames = new Set(Object.values(renames));

  for (const change of changes) {
    if (change.type === "added") {
      // If it's a rename target, that's fine
      if (renameNewNames.has(change.field)) continue;
      // New field must not be required without a default
      // (this check would need the actual field info — done at table level)
    } else if (change.type === "removed") {
      // If it's a rename source, that's fine
      if (renameOldNames.has(change.field)) continue;
      // Removing fields is auto-migratable
    } else if (change.type === "typeChanged") {
      if (!hasTransform) {
        errors.push(
          `Field "${change.field}" changed type from "${change.oldType}" to "${change.newType}" ` +
            `but no transform function provided in migration version ${targetVersion}`
        );
      }
    }
  }

  return errors;
}

File: internal/schema/diff.go
package schema

import "fmt"

// SchemaChange describes a difference between two schema versions.
type SchemaChange struct {
	Type    string // "added", "removed", "typeChanged"
	Field   string
	OldType string
	NewType string
}

// DiffSchemas compares a stored schema with the current compiled schema.
func DiffSchemas(stored *StoredSchema, current *CompiledSchema) []SchemaChange {
	var changes []SchemaChange

	storedMap := make(map[string]*StoredColumnDef)
	for i := range stored.Columns {
		storedMap[stored.Columns[i].Name] = &stored.Columns[i]
	}

	currentMap := make(map[string]*CompiledField)
	for i := range current.Fields {
		currentMap[current.Fields[i].Name] = &current.Fields[i]
	}

	// Removed fields
	for name := range storedMap {
		if _, ok := currentMap[name]; !ok {
			changes = append(changes, SchemaChange{Type: "removed", Field: name})
		}
	}

	// Added or changed fields
	for name, field := range currentMap {
		old, exists := storedMap[name]
		if !exists {
			changes = append(changes, SchemaChange{Type: "added", Field: name})
		} else if old.Type != string(field.Kind) {
			changes = append(changes, SchemaChange{
				Type:    "typeChanged",
				Field:   name,
				OldType: old.Type,
				NewType: string(field.Kind),
			})
		}
	}

	return changes
}

// SchemasEqual checks if a stored schema matches the current compiled schema.
func SchemasEqual(stored *StoredSchema, current *CompiledSchema) bool {
	if len(stored.Columns) != len(current.Fields) {
		return false
	}
	for i := range stored.Columns {
		if stored.Columns[i].Name != current.Fields[i].Name {
			return false
		}
		if stored.Columns[i].Type != string(current.Fields[i].Kind) {
			return false
		}
	}
	return true
}

// ValidateMigration checks schema changes against provided migration steps.
func ValidateMigration(changes []SchemaChange, migrations []MigrationStep, targetVersion int) []string {
	var errors []string

	var migration *MigrationStep
	for i := range migrations {
		if migrations[i].Version == targetVersion {
			migration = &migrations[i]
			break
		}
	}

	renames := make(map[string]string)
	hasTransform := false
	if migration != nil {
		renames = migration.Rename
		// Transform is handled on JS side; we just check if migration exists
		hasTransform = true
	}

	renameOldNames := make(map[string]bool)
	renameNewNames := make(map[string]bool)
	for old, new_ := range renames {
		renameOldNames[old] = true
		renameNewNames[new_] = true
	}

	for _, change := range changes {
		switch change.Type {
		case "added":
			if renameNewNames[change.Field] {
				continue
			}
		case "removed":
			if renameOldNames[change.Field] {
				continue
			}
		case "typeChanged":
			if !hasTransform {
				errors = append(errors, fmt.Sprintf(
					`Field "%s" changed type from "%s" to "%s" but no transform function provided in migration version %d`,
					change.Field, change.OldType, change.NewType, targetVersion,
				))
			}
		}
	}

	return errors
}
